subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyHandler.java

```diff
@@ -127,10 +129,24 @@ private Dependency doAdd(Configuration configuration, Object dependencyNotation,
             configuration.extendsFrom(other);
             return null;
         }
-
-        Dependency dependency = create(dependencyNotation, configureClosure);
-        configuration.getDependencies().add(dependency);
-        return dependency;
+        if (dependencyNotation instanceof Provider<?>) {
+            Provider<?> lazy = (Provider<?>) dependencyNotation;
+
+            Provider<Dependency> lazyDependency = lazy.map(lazyNotation -> {
+                if (lazy instanceof HasConfigurableValue) {
+                    HasConfigurableValue configurableLazy = (HasConfigurableValue) lazy;
+                    configurableLazy.disallowChanges();
```

Comment:
@adammurdoch Is this a reasonable use of this method (`disallowChanges`)?

There are no other places in production code I could detect it being used, it's only ever used in tests.

Part of me wishes that this method had a `String reason` argument to explain why in the eventual error generated.

Reply:
What are you aiming to do here? Prevent changes to the value of the provider after the set of dependencies has been queried?

If so, then there are a few issues. Firstly, `disallowChanges()` doesn't prevent the value from changing, it prevents changes to the wiring instead and it's not something a consumer should ever call. Finalizing the value would be a better option. Secondly, the timing is inconsistent with other places. We shouldn't finalise the value of an upstream dependency provider until the set of dependencies is finalised, not when the dependencies are queried. I wouldn't finalise some of the values but leave some mutable. Currently, finalisation happens when the configuration is involved in dependency resolution. When it _should_ happen is another story, but I would not try to change that in this PR. Finally, I probably wouldn't implement an ad hoc solution here, but would instead continue to improve the infrastructure so that this is a cross-cutting thing that containers + providers just do.

Reply:
> What are you aiming to do here?

Satisfy the request from @melix here:

> show that mutation after dependencies have been queried provide a reasonable error message

@melix Is it fine if I just completely remove the locking behaviour and the associated tests. I don't think there's much value-added here anyways.

Reply:
I would either remove it for now, or replace it with some cross-cutting solutions that work more generally. I wouldn't do some ad hoc thing here that solves the problem in a partial and inconsistent way (eg if I query the dependency set I can't change the upstream _property_, but I can still remove the dependency object from the set, or add more dependencies, or mutate the dependency object itself, and mutate the other dependency objects added in some non-lazy way). 

I think it's fine to not add anything special here for now and for us to later (possibly immediately after this PR) add some cross-cutting stuff to use here and everywhere else.